policy/composer.go

```diff
@@ -362,3 +474,42 @@ func isOptionalNone(e ast.Expr) bool {
 		e.AsCall().FunctionName() == "optional.none" &&
 		len(e.AsCall().Args()) == 0
 }
+
+func removeIneligibleConditions(e ast.NavigableExpr, conditionMap map[int64]int) {
+	for _, id := range comprehensionSubExprIDs(e) {
+		if _, found := conditionMap[id]; found {
+			delete(conditionMap, id)
+		}
+	}
+}
+
+func comprehensionSubExprIDs(e ast.NavigableExpr) []int64 {
+	compre := e.AsComprehension()
+	// Almost the same as e.Children(), but skips the iteration range
+	return enumerateExprIDs(
+		compre.AccuInit().(ast.NavigableExpr),
+		compre.LoopCondition().(ast.NavigableExpr),
+		compre.LoopStep().(ast.NavigableExpr),
+		compre.Result().(ast.NavigableExpr),
```

Comment:
Do you think we can a test case with nested macros with a deeply nested expr inside? Anecdotally, this is where I ran into the most trouble with CSE, [so I ended up creating some baseline cases for it](https://github.com/google/cel-java/blob/a9c96754f6852ad9818a9cbe58c71a4d165be897/optimizer/src/test/resources/subexpression_unparsed.baseline#L321-L367). For example, I'm not 100% confident that we should be including the loop step / result here.

Reply:
Since the loop step and result can access internal state, e.g. `@result`, there's some possibility that a rewrite of any expression apart from the range within a comprehension would be problematic. I'm happy to add the baselines though!

Reply:
Since we're not mangling expression identifiers within comprehensions, I'm going to continue with the more conservative approach and omit all comprehension sub-expressions.